### PR TITLE
Put Azul behind AWS WAF (#4000)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Connected issue: #4000

### Author

- [x] Target branch is `develop`
- [x] Source branch is `issues/<github-handle>/<issue#>-<slug>`
- [x] PR title references the connected issue
- [x] PR title matches<sup>1</sup> title of connected issue         <sub>or comment in PR explains why they're different</sub>
- [x] Title of at least one commit references connected issue
- [x] PR is connected to issue via Zenhub 
- [x] PR description links to connected issue
- [x] Added `partial` label to PR                                    <sub>or this PR completely resolves the connected issue</sub>

<sup>1</sup> when the issue title describes a problem, the PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain                <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or there is no temporary hotfix for the connected issue on the `prod` branch</sub>

### Author (requirements)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after rejection)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `demo` or `no demo`
- [x] Commented on connected issue about demo expectations          <sub>or labelled connected issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Ticket is in *Review requested* column
- [x] Requested review from peer reviewer
- [x] Assigned PR to peer reviewer

### Peer reviewer (after rejection)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] `N reviews` label on PR reflects prior peer reviews with changes requested
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or connected issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Deploy PR branch to GitLab to update permission policy
- [x] Branch pushed to GitLab and added `sandbox` label             <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                                       <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issue to Merged column
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [ ] Shortened the PR chain                                        <sub>or this PR is not labeled `chain`</sub>
- [ ] Pushed merge commit to GitLab                                 <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from GitHub and GitLab
- [ ] Build passes on GitLab

### Operator (reindex) 

- [ ] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [ ] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures                     <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment                      <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Shouldn't this be chained to #4028? I thought a regional gateway was a prerequisite for placing it behind WAF. If my assumption is wrong, I'd like to know. Please also prepare a pre-review demo of this in a PL slot.

**Comment by danielsotirhos** [comment]
> I thought a regional gateway was a prerequisite for placing it behind WAF. If my assumption is wrong, I'd like to know.

The Web ACLs must themselves be `regional`, however I do not see a limitation that they can only be attached to regional API Gateways.

https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-control-access-aws-waf.html

Screen Shot 2022-04-18 at 10 45 55 AM

---

As demonstrated here, I have an `Edge` API Gateway:

Screen Shot 2022-04-18 at 10 38 42 AM

And when creating an ACL...:

Screen Shot 2022-04-18 at 10 42 59 AM

...I am able to select my `Edge` API Gateway

Screen Shot 2022-04-18 at 10 44 57 AM

Screen Shot 2022-04-18 at 10 48 16 AM

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
From my previous review

> Please also prepare a pre-review demo of this in a PL slot.

by "this" I meant the entire functionality that this PR adds.

File: terraform/api_gateway.tf.json.template.py [hunk 1]
```diff
@@ -93,167 +93,202 @@ def for_domain(cls, domain):
...
+        # Chalice doesn't expose the ARN of the API Gateway stages, so we
+        # construct the ARN manually using this workaround.
+        # https://github.com/aws/chalice/issues/1816#issuecomment-1012231084
+        {
+            "aws_wafv2_web_acl_association": {
+                f"azul_{lambda_}_waf": {
+                    "resource_arn": f"${{module.chalice_{lambda_}.rest_api_arn}}"
+                                    f"/stages/${{module.chalice_{lambda_}.stage_name}}",
+                    "web_acl_arn": "${aws_wafv2_web_acl.azul_waf_web_acl.arn}"
+                } for lambda_ in ('indexer', 'service')
```
Comment (hannes-ucsc) [inline terraform/api_gateway.tf.json.template.py:290]: There already is a comprehension for this. Why did you roll your own? Using the existing comprehension would eliminate the need to convert it to a generator expressions.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
It appears that the [Fix pep8 warnings](https://github.com/DataBiosphere/azul/pull/4084/commits/0a65ae3a206df4a1dfc68d0c7776c442dcba6b4b) commit is fixing a regression introduced by the previous commit, which you authored. You shouldn't introduce a violation in one commit and fix it in another commit of the same PR. So either squash that commit out or move it to the beginning of the history (in case the pep8 violation is already on develop).

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
The fact that we can use the same ACL for both WAFs is probably incidental. We should create a separate ACL for each. And don't fix the line length violations in the existing code. I don't particularly like the fixes, and they exaggerate the diff, at least when taking whitespace into account.

File: terraform/api_gateway.tf.json.template.py [hunk 2]
```diff
@@ -253,6 +253,32 @@ def for_domain(cls, domain):
                     "policy": lambda_.policy,
                     "role": "${aws_iam_role.%s.id}" % lambda_.name
                 },
+            },
+            # Chalice doesn't expose the ARN of the API Gateway stages, so we
+            # construct the ARN manually using this workaround.
+            # https://github.com/aws/chalice/issues/1816#issuecomment-1012231084
+            "aws_wafv2_web_acl_association": {
+                f"azul_{lambda_.name}_waf": {
```
Comment (hannes-ucsc) [inline terraform/api_gateway.tf.json.template.py:261]:
```suggestion
                lambda_.name: {
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Review by achave11 — APPROVED** [review]

**Comment by achave11** [comment]
PR failed `deploy` job in [sandbox](https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/45788).

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 3]
```diff
@@ -502,28 +502,10 @@ def remove_inconsequential_statements(statements: list[JSON]) -> list[JSON]:
...
-                                             LogGroupName=log_group_name,
-                                             LogStream='*',
-                                             LogStreamName='*')
-                            for log_group_name in [
-                                '/aws/apigateway/azul-*',
-                                '/aws/lambda/azul-*',
-                                '/aws/aes/domains/azul-*'
-                            ]
-                        )
+                        'actions': ['logs:*'],
```
Comment (hannes-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py:507]: https://github.com/DataBiosphere/azul/issues/4207

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Please change title of commit to  `Consolidate CW logs permissions in GitLab boundary` and add a FIXME referring to https://github.com/DataBiosphere/azul/issues/4207

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 4]
```diff
@@ -502,28 +502,12 @@ def remove_inconsequential_statements(statements: list[JSON]) -> list[JSON]:
                         'resources': ['*']
                     },
 
-                    allow_global_actions('CloudWatch Logs'),
+                    # CloudWatch Logs
+                    # FIXME: Tighten GitLab security boundary
```
Comment (hannes-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py:506]: Commit title should reference FIXME.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**